cpp/include/cudf/detail/copy_if.cuh

```diff
@@ -0,0 +1,412 @@
+
+    cudf::size_type per_thread =
+      elements_per_thread(scatter, input.size(), block_size);
+    cudf::util::cuda::grid_config_1d grid{input.size(),
+                                          block_size, per_thread};
+
+    rmm::device_scalar<cudf::size_type> null_count{0, stream, mr};
+    if (has_valid) {
+      // Have to initialize the output mask to all zeros because we may update
+      // it with atomicOr().
+      CUDA_TRY(cudaMemsetAsync(static_cast<void*>(output.null_mask()), 0,
+                               cudf::bitmask_allocation_size_bytes(output.size()),
+                               stream));
+    }
+    
+    auto input_device_view  = cudf::column_device_view::create(input, stream);
+    scatter<<<grid.num_blocks, block_size, 0, stream>>>
+      (output.data<T>(), output.null_mask(), null_count.data(),
+       *input_device_view, block_offsets, input.size(), per_thread, filter);
+
+    if (has_valid) {
+      output.set_null_count(null_count.value());
+    }
+    return output_column;
+  }
+
+};
+} // namespace
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Filters `input` using a Filter function object
+ * 
+ * @p filter must be a functor or lambda with the following signature:
+ * __device__ bool operator()(cudf::size_type i);
+ * It will return true if element i of @p input should be copied, 
+ * false otherwise.
+ *
+ * @tparam Filter the filter functor type
+ * @param[in] input The table_view to filter
+ * @param[in] filter A function object that takes an index and returns a bool
+ * @return unique_ptr<table> The table generated from filtered `input`.
+ */
+template <typename Filter>
+std::unique_ptr<experimental::table> copy_if(table_view const& input, Filter filter,
+                          rmm::mr::device_memory_resource *mr =
+                              rmm::mr::get_default_resource(),
+                          cudaStream_t stream = 0) {
+
+    std::vector<std::unique_ptr<column>> out_columns(input.num_columns());
+    if (0 == input.num_rows() || 0 == input.num_columns()) {
+        std::transform(input.begin(), input.end(), out_columns.begin(),
+                [&stream] (auto col_view){return detail::empty_like(col_view, stream);});
+
+        return std::make_unique<experimental::table>(std::move(out_columns));
+    }
+
+    constexpr int block_size = 256;
+    cudf::size_type per_thread =
+      elements_per_thread(compute_block_counts<Filter, block_size>,
+                          input.num_rows(), block_size);
+    cudf::util::cuda::grid_config_1d grid{input.num_rows(), block_size, per_thread};
+
+    // allocate temp storage for block counts and offsets
+    // TODO: use an uninitialized buffer to avoid the initialization kernel
+    rmm::device_vector<cudf::size_type> temp_counts(2 * grid.num_blocks);
+    cudf::size_type *block_counts = thrust::raw_pointer_cast(temp_counts.data());
+    cudf::size_type *block_offsets = block_counts + grid.num_blocks;
```

Comment:
This can't result in misaligned memory accesses AFAIK, they are just ints, not structs of a weird size.

I wrote this code. It was originally a direct RMM_ALLOC, so it simplified the code and saved an alloc to do it this way.